---
title: "DADA2 Workflow"
author: "Yining Sun"
date: "2023-04-17"
output: html_document
editor_options: 
  chunk_output_type: console
---


# Let's strat! Set workdir, source code and load packages!
```{r prep-to-start}
# be in the right place
setwd("/local/workdir/ys636/BIOMI6300-Project2")

# source functions.R
source("/local/workdir/ys636/BIOMI6300-Project2/code/functions.R")

# load packages with pacman
pacman::p_load(dada2, tidyverse, phyloseq, patchwork, Biostrings, install = FALSE)

# check dada2 version
packageVersion("dada2")
```


# Set path
```{r set-path}

# set paths for meat and dairy folders of gzipped files
path_meat <- ("/local/workdir/ys636/BIOMI6300-Project2/data/meat")
path_dairy <- ("/local/workdir/ys636/BIOMI6300-Project2/data/dairy")

# check path
list.files(path_meat)  # 16 fastq.gz files, looks good!
list.files(path_dairy)  # 16 fastq.gz files, looks good, too!

```

# Create variables for forward and reverse reads
```{r F-and-R-variables}
#R1 is forward, R2 is reverse

# F and R variables for MEAT group
F_reads_meat <- sort(list.files(path_meat, pattern = "_1.fastq.gz",
                                 full.names = TRUE))

R_reads_meat <- sort(list.files(path_meat, pattern = "_2.fastq.gz",
                                 full.names = TRUE))

# F and R variables for Dairy group
F_reads_dairy <- sort(list.files(path_dairy, pattern = "_1.fastq.gz",
                                 full.names = TRUE))

R_reads_dairy <- sort(list.files(path_dairy, pattern = "_2.fastq.gz",
                                 full.names = TRUE))
```


# Quality plot of raw reads
```{r QualPlot-untrimmed}
# show the quality of each base on the reads of all the 8 samples in both groups

# Meat group
## Forward
F_Qual_plot_meat <- plotQualityProfile(F_reads_meat)
F_Qual_plot_meat                      ### The first 190-200 bases had quality scores over 20
## Reverse
R_Qual_plot_meat <- plotQualityProfile(R_reads_meat)
R_Qual_plot_meat                      ### The first 160 bases had quality scores over 20

# Dairy group
## Forward
F_Qual_plot_dairy <- plotQualityProfile(F_reads_dairy)
F_Qual_plot_dairy             
### The last 4 samples' first 200-210 bases had quality scores over 20, the first 4 samppels looked good with all reads

## Reverse
R_Qual_plot_dairy <- plotQualityProfile(R_reads_dairy)
R_Qual_plot_dairy             
### The last 4 samples' first 160 bases had quality scores over 20, the first 4 samppels looked good with all reads

### based on the quality plots of all the 16 samples from both groups, I decided to set the quality score cutoff as 20, rather than 30.

```


# Modify file names for the forward and reverse filtered reads
```{r variables-trimmed-reads}

# Create characters for trimmed reads 
sample_meat <- scan(file = "/local/workdir/ys636/BIOMI6300-Project2/data/meat.txt", what = "character")
sample_meat

sample_dairy <- scan(file = "/local/workdir/ys636/BIOMI6300-Project2/data/dairy.txt", what = "character")
sample_dairy

# create variables to hold file names
## Meat group
filtered_F_reads_meat <- file.path(path_meat, "filtered", paste0(sample_meat, "_1_filtered.fastq.gz"))
filtered_R_reads_meat <- file.path(path_meat, "filtered", paste0(sample_meat, "_2_filtered.fastq.gz"))

## Dairy group
filtered_F_reads_dairy <- file.path(path_dairy, "filtered", paste0(sample_dairy, "_1_filtered.fastq.gz"))
filtered_R_reads_dairy <- file.path(path_dairy, "filtered", paste0(sample_dairy, "_2_filtered.fastq.gz"))

```


# Filter and trim the raw reads!!
```{r filter-and-trim}

# Meat group
filtered_out_meat <- filterAndTrim(F_reads_meat, filtered_F_reads_meat, 
                              R_reads_meat, filtered_R_reads_meat,
                              truncLen = 0, trimRight = 100,
                              maxN = 0, truncQ = 2, 
                              rm.phix = TRUE, compress = TRUE, 
                              multithread = TRUE)
## kept the first 200 bases in forward reads and the first 200 bases in reverse reads


# Dairy group
filtered_out_dairy <- filterAndTrim(F_reads_dairy, filtered_F_reads_dairy, 
                              R_reads_dairy, filtered_R_reads_dairy,
                              trimRight = 50,
                              maxN = 0, truncQ = 2, 
                              rm.phix = TRUE, compress = TRUE, 
                              multithread = TRUE)
## kept the first 250 bases for both forward and reverse reads


```


# Plot the quality of filtered and trimmed reads for both groups
```{r QualPlot-trim}

# Meat group
## forward
filtered_F_Qual_plot_meat <- plotQualityProfile(filtered_F_reads_meat)
filtered_F_Qual_plot_meat    
## reverse
filtered_R_Qual_plot_meat <- plotQualityProfile(filtered_R_reads_meat)
filtered_R_Qual_plot_meat  


# Dairy group
## forward
filtered_F_Qual_plot_dairy <- plotQualityProfile(filtered_F_reads_dairy)
filtered_F_Qual_plot_dairy    
## reverse
filtered_R_Qual_plot_dairy <- plotQualityProfile(filtered_R_reads_dairy)
filtered_R_Qual_plot_dairy  


### After this step, the reads for each sample in both groups looked nice and neat - all bases' quality score over 20

### Proceed to learn errors

```

# Put QualPlots of before and after filterAndtrim together
```{r QualPlot-before-and-after-trim}
# Meat group
(F_Qual_plot_meat+R_Qual_plot_meat)/(filtered_F_Qual_plot_meat+filtered_R_Qual_plot_meat)

# Dairy group
(F_Qual_plot_dairy+R_Qual_plot_dairy)/(filtered_F_Qual_plot_dairy+filtered_R_Qual_plot_dairy)

```


# Error model
```{r learn-errors}

# Meat group
err_forward_reads_meat <- learnErrors(filtered_F_reads_meat, multithread = TRUE)
err_reverse_reads_meat <- learnErrors(filtered_R_reads_meat, multithread = TRUE)

# Dairy group
err_forward_reads_dairy <- learnErrors(filtered_F_reads_dairy, multithread = TRUE)
err_reverse_reads_dairy <- learnErrors(filtered_R_reads_dairy, multithread = TRUE)


# plot the errors
# Meat group
plotErrors(err_forward_reads_meat, nominalQ = TRUE)
plotErrors(err_reverse_reads_meat, nominalQ = TRUE)

# Dairy group
plotErrors(err_forward_reads_dairy, nominalQ = TRUE)
plotErrors(err_reverse_reads_dairy, nominalQ = TRUE)

```


# Infer ASVs on the forward and reverse seqs with DADA2
```{r infer-ASVs}
# Meat group
## run dada2 on the forward reads
dada_F_meat <- dada(filtered_F_reads_meat, err = err_forward_reads_meat, multithread =  TRUE)
dada_F_meat
## run dada2 on the reverse reads
dada_R_meat <- dada(filtered_R_reads_meat, err = err_reverse_reads_meat, multithread =  TRUE)
dada_R_meat

### For each sample in the meat group, they all had over 100 sequence variants inferred.

# Dairy group
## run dada2 on the forward reads
dada_F_dairy <- dada(filtered_F_reads_dairy, err = err_forward_reads_dairy, multithread =  TRUE)
dada_F_dairy
## run dada2 on the reverse reads
dada_R_dairy <- dada(filtered_R_reads_dairy, err = err_reverse_reads_dairy, multithread =  TRUE)
dada_R_dairy

### For sample No.2,5,6,8 in the dairy group, they only had less than 20 sequence variants inferred, which was surprising, especially compared to numbers of corresponding forward reads.

```


# Merge forward and reverse ASVs
```{r merge-FandR-ASVs}

# Meat group
merged_amplicons_meat <- mergePairs(dada_F_meat, filtered_F_reads_meat,
                               dada_R_meat, filtered_R_reads_meat,
                               verbose = TRUE, justConcatenate = TRUE)

# Dairy group
merged_amplicons_dairy <- mergePairs(dada_F_dairy, filtered_F_reads_dairy,
                               dada_R_dairy, filtered_R_reads_dairy,
                               verbose = TRUE, justConcatenate = TRUE)


### Attention!!!
### added 'justConcatenate = TRUE' in order to have more merged pairs. 
### If not adding this condition, there will be no or few merged pairs for later analysis. 
### reason: this study sequenced V3 & V4 regions, not only one V region
### so the sequences could be longer than those we did in class.
### Will keep an eye for this in later analysis!


## Evaluate the output
# typeof(merged_amplicons_meat)
# merged_amplicons_meat[1]
# length(merged_amplicons_meat)
# names(merged_amplicons_meat)


```


# Generate a count table
```{r count-tab}
# Meat group
seqtab_meat <- makeSequenceTable(merged_amplicons_meat)
class(seqtab_meat) ### "matrix" "array"
typeof(seqtab_meat) ### "integer"
dim(seqtab_meat) ### 8 rows, 28388 columns
View(seqtab_meat)


# Dairy group
seqtab_dairy <- makeSequenceTable(merged_amplicons_dairy)
class(seqtab_dairy) ### "matrix" "array"
typeof(seqtab_dairy) ### "integer"
dim(seqtab_dairy) ### 8 rows, 25695 rows
View(seqtab_dairy)


### for both groups' seqtabs, each sequence only exists in one sample.
### for example, "1000" reads of sequence1 in sample1, but 0" reads in sample2-8.

```






